### PR TITLE
tools: add configuration.bios.settings to the preamble

### PR DESCRIPTION
**Pull request by ldorau**
Requires:
- [x] #1178 

---
This change is [Reviewable](https://reviewable.io/reviews/pmem/rpma/1179)

### CODE REVIEW

**Review by janekmi — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/1179)* status: all files reviewed, 2 unresolved discussions (waiting on @ldorau)

___
*[tools/perf/config.json.example, line 50 at r1](https://reviewable.io/reviews/pmem/rpma/1179#-MfnJwgYAeMPX9byA-nE:-MfnJwgYAeMPX9byA-nF:b-b5lobe) ([raw file](https://github.com/pmem/rpma/blob/806cd823f4b904d40048bfd5ce06cf632090004f/tools/perf/config.json.example#L50)):*
> ```
>             "bios": {
>                 "settings" : {
>                     "type": "kvtable",
> ```

I think this is convenient when we have complete control over files (the end-users are not intended to edit them).
In the case of files fully provided by the end-users I thought about forcing a structure (which is tricky enough) but not forcing the syntax.

Long story short: I think we can add this key:value pair automatically so the end-user does not have to remember to set it and set it right. What do you think?
___
*[tools/perf/lib/Report.py, line 24 at r1](https://reviewable.io/reviews/pmem/rpma/1179#-MfnLv2S4DDD2rpdTHqe:-MfnLv2S4DDD2rpdTHqf:b-efxamg) ([raw file](https://github.com/pmem/rpma/blob/806cd823f4b904d40048bfd5ce06cf632090004f/tools/perf/lib/Report.py#L24)):*
> ```Python
> 
>         preamble = Part(loader, env, 'preamble')
>         preamble._process_variables_level(variables, {})
> ```

Functions prefixed with `_` are internal by convention. If you want to use it, you have to remove the`_`.

**Review by janekmi — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/1179)* status: all files reviewed, 2 unresolved discussions (waiting on @ldorau)

___
*[tools/perf/lib/Report.py, line 24 at r1](https://reviewable.io/reviews/pmem/rpma/1179#-MfnLv2S4DDD2rpdTHqe:-MfnN4Ao-mh6Krcd7K79:b-977qmi) ([raw file](https://github.com/pmem/rpma/blob/806cd823f4b904d40048bfd5ce06cf632090004f/tools/perf/lib/Report.py#L24)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

Functions prefixed with `_` are internal by convention. If you want to use it, you have to remove the`_`.
</blockquote></details>

My bad. Private methods should be prefixed with `__` (double underscore). It requires a fix.

Ref: https://www.geeksforgeeks.org/private-methods-in-python/

**Review by ldorau — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/1179)* status: 2 of 4 files reviewed, 2 unresolved discussions (waiting on @janekmi)

___
*[tools/perf/config.json.example, line 50 at r1](https://reviewable.io/reviews/pmem/rpma/1179#-MfnJwgYAeMPX9byA-nE:-Mfpu148D-oFkgbRZGaJ:b-ol7rx1) ([raw file](https://github.com/pmem/rpma/blob/806cd823f4b904d40048bfd5ce06cf632090004f/tools/perf/config.json.example#L50)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

I think this is convenient when we have complete control over files (the end-users are not intended to edit them).
In the case of files fully provided by the end-users I thought about forcing a structure (which is tricky enough) but not forcing the syntax.

Long story short: I think we can add this key:value pair automatically so the end-user does not have to remember to set it and set it right. What do you think?
</blockquote></details>

Where and when add it automatically? It is used by `_process_variables_level`:
```
if v["type"] == "kvtable":
    variables[k] = dict2kvtable(v, common)
else:
    raise ValueError
```
___
*[tools/perf/lib/Report.py, line 24 at r1](https://reviewable.io/reviews/pmem/rpma/1179#-MfnLv2S4DDD2rpdTHqe:-Mfpw46gB5x9nPDXLeD9:b-896fix) ([raw file](https://github.com/pmem/rpma/blob/806cd823f4b904d40048bfd5ce06cf632090004f/tools/perf/lib/Report.py#L24)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

My bad. Private methods should be prefixed with `__` (double underscore). It requires a fix.

Ref: https://www.geeksforgeeks.org/private-methods-in-python/
</blockquote></details>

Done.

**Review by ldorau — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/1179)* status: 1 of 4 files reviewed, 2 unresolved discussions (waiting on @janekmi)

___
*[tools/perf/config.json.example, line 50 at r1](https://reviewable.io/reviews/pmem/rpma/1179#-MfnJwgYAeMPX9byA-nE:-Mfq364J2txFS1jCuABO:b65rwd4) ([raw file](https://github.com/pmem/rpma/blob/806cd823f4b904d40048bfd5ce06cf632090004f/tools/perf/config.json.example#L50)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

Where and when add it automatically? It is used by `_process_variables_level`:
```
if v["type"] == "kvtable":
    variables[k] = dict2kvtable(v, common)
else:
    raise ValueError
```
</blockquote></details>

Ah, I know. Done.

**Review by ldorau — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/1179)* status: 1 of 4 files reviewed, 2 unresolved discussions (waiting on @janekmi)

___
*[tools/perf/config.json.example, line 50 at r1](https://reviewable.io/reviews/pmem/rpma/1179#-MfnJwgYAeMPX9byA-nE:-Mfq4DCYCNBdf-FY2Uj0:b-896fix) ([raw file](https://github.com/pmem/rpma/blob/806cd823f4b904d40048bfd5ce06cf632090004f/tools/perf/config.json.example#L50)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

Ah, I know. Done.
</blockquote></details>

Done.

**Review by osalyk — COMMENTED** [review]
> Reviewed 1 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/1179)* status: 2 of 4 files reviewed, 2 unresolved discussions (waiting on @janekmi)

**Review by janekmi — CHANGES_REQUESTED** [review]
> Reviewed 1 of 3 files at r2, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/1179)* status: all files reviewed, 2 unresolved discussions (waiting on @ldorau)

___
*[tools/perf/lib/Report.py, line 26 at r3](https://reviewable.io/reviews/pmem/rpma/1179#-MfqGjUU-DDdHISmBkaw:-MfqGjUU-DDdHISmBkax:b-v69cld) ([raw file](https://github.com/pmem/rpma/blob/e0c0d97404bfd5ae7fb19812b4910a247c93671d/tools/perf/lib/Report.py#L26)):*
> ```Python
>                 if 'settings' in variables['configuration']['bios']:
>                     if 'type' not in variables['configuration']['bios']['settings']:
>                         variables['configuration']['bios']['settings']['type'] = 'kvtable'
> ```

```python
# XXX type validation is missing
if 'configuration' not in variables:
    raise SyntaxError('config.json misses ['report']['configuration'] entry')
else if 'bios' not in variables['configuration']:
    raise SyntaxError('config.json misses ['report']['configuration']['bios'] entry')
else if 'settings' not in variables['configuration']['bios']:
    raise SyntaxError('config.json misses ['report']['configuration']['bios']['settings'] entry')

# the only correct type is 'kvtable'
variables['configuration']['bios']['settings']['type'] = 'kvtable'
```

And I'm not sure whether `SyntaxErrror` is the right one. We can reconsider it later.
___
*[tools/perf/lib/Report.py, line 26 at r3](https://reviewable.io/reviews/pmem/rpma/1179#-MfqIjGJ1Xo2P2eiIBkk:-MfqIjGJ1Xo2P2eiIBkl:b8zl95q) ([raw file](https://github.com/pmem/rpma/blob/e0c0d97404bfd5ae7fb19812b4910a247c93671d/tools/perf/lib/Report.py#L26)):*
> ```Python
>                 if 'settings' in variables['configuration']['bios']:
>                     if 'type' not in variables['configuration']['bios']['settings']:
>                         variables['configuration']['bios']['settings']['type'] = 'kvtable'
> ```

TBH I think this part will become quickly a separate function performing the configuration validation. Some parts are optional some are not.

**Review by ldorau — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/1179)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @janekmi)

___
*[tools/perf/lib/Report.py, line 26 at r3](https://reviewable.io/reviews/pmem/rpma/1179#-MfqGjUU-DDdHISmBkaw:-MfqTtre1eNHBnPDRSaT:b-896fix) ([raw file](https://github.com/pmem/rpma/blob/e0c0d97404bfd5ae7fb19812b4910a247c93671d/tools/perf/lib/Report.py#L26)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

```python
# XXX type validation is missing
if 'configuration' not in variables:
    raise SyntaxError('config.json misses ['report']['configuration'] entry')
else if 'bios' not in variables['configuration']:
    raise SyntaxError('config.json misses ['report']['configuration']['bios'] entry')
else if 'settings' not in variables['configuration']['bios']:
    raise SyntaxError('config.json misses ['report']['configuration']['bios']['settings'] entry')

# the only correct type is 'kvtable'
variables['configuration']['bios']['settings']['type'] = 'kvtable'
```

And I'm not sure whether `SyntaxErrror` is the right one. We can reconsider it later.
</blockquote></details>

Done.

**Review by janekmi — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/1179)* status: all files reviewed, 1 unresolved discussion (waiting on @ldorau)

___
*[tools/perf/lib/Report.py, line 26 at r3](https://reviewable.io/reviews/pmem/rpma/1179#-MfqGjUU-DDdHISmBkaw:-MfqUwxf8bR9GIIqy-4R:bv1e0tj) ([raw file](https://github.com/pmem/rpma/blob/e0c0d97404bfd5ae7fb19812b4910a247c93671d/tools/perf/lib/Report.py#L26)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I'm sure you haven't tested this. :rofl:

**Review by ldorau — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/1179)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @janekmi)

___
*[tools/perf/lib/Report.py, line 26 at r3](https://reviewable.io/reviews/pmem/rpma/1179#-MfqGjUU-DDdHISmBkaw:-MfqVGmK9GwVOpwZ2-uC:b-896fix) ([raw file](https://github.com/pmem/rpma/blob/e0c0d97404bfd5ae7fb19812b4910a247c93671d/tools/perf/lib/Report.py#L26)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

I'm sure you haven't tested this. :rofl: 
</blockquote></details>

Done.

**Review by osalyk — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/1179#-:-MfqVUFJA2MCqGDFS8UW:bnfp4nl)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @janekmi)

**Review by janekmi — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/pmem/rpma/1179#-:-MfqVv3T9MtEj514y2cy:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @ldorau)
